### DISCUSSION TITLE
function envelope(dsyn, dobs)

### DISCUSSION THREAD

**Discussion by venividivici666** [slimgroup/JUDI.jl#236] [Q&A]
Hello! When I try "FWI with user provided misfit function", the function envelope(dsyn, dobs) shows "MethodError: Cannot `convert` an object of type judiVector{Float32, Matrix{Float32}} to an object of type ComplexF32", how can I fix it? Thanks a lot.

**Comment by mloubout** [comment]
Wich example/script are you running?

The example in `examples/scripts` is tested as part of CI and so is the notebook. If you are running a script you are running yourself could you share a minimal script that reproduces the issue.

**Reply by venividivici666** [reply]
using FFTW, Zygote

n2(x) = x / norm(x, 2)

function Hilbert(x)
    n = size(x, 1)
    σ = sign.(-n/2+1:n/2)
    y = imag(ifft(fftshift(σ.*fftshift(fft(x, 1), 1), 1), 1))
    return y
end

HLoss(dsyn, dobs) = sum(abs2.((dsyn - dobs) .+ 1im .* Hilbert(dsyn - dobs)))

function envelope(dsyn, dobs)
    ϕ = HLoss(n2(dsyn), n2(dobs))
    g = gradient(xs -> HLoss(n2(xs), n2(dobs)), dsyn)
    # Zygote always returns a tuple, in this case (g,)
    return ϕ, real.(g[1])
end

I run the examples/scripts/fwi_example_studentst.jl using the envelope misfit function above

**Reply by mloubout** [reply]
` examples/scripts/fwi_example_studentst.jl` runs fine, where do you get the error?

**Reply by venividivici666** [reply]
examples/scripts/fwi_example_studentst.jl runs fine, but when I want to use the envelope misfit function, it shows "MethodError: Cannot convert an object of type judiVector{Float32, Matrix{Float32}} to an object of type ComplexF32"

**Reply by mloubout** [reply]
The script runs fine using envelope instead of a student's misfit as well.

I would be happy to assist, but I need a reproducible error script as these features are tested and working in the examples.

Cheers

**Reply by venividivici666** [reply]
![image](https://github.com/slimgroup/JUDI.jl/assets/60508154/0b0cf097-3176-481b-8f8d-bde2791198c4)

**Reply by venividivici666** [reply]
And before that is the examples/scripts/fwi_example_studentst.jl

**Reply by mloubout** [reply]
That's a different error than what you pointed out earlier, this just says you are calling `objective_function` incorrectly as it expects `objective_function(x, misfit)` and you are doing `objective_function(x, d_obs, misfit)`, which isn't defined

**Reply by venividivici666** [reply]
When I fix the objective_function, another problem is below:
![Uploading image.png…]()

**Comment by venividivici666** [comment]
![image](https://github.com/slimgroup/JUDI.jl/assets/60508154/86ceb8f1-045e-4d1a-92fa-44a78595cd22)

**Reply by mloubout** [reply]
Please check the tutorial

https://slimgroup.github.io/JUDI.jl/dev/tutorials/05_custom_misfit/

https://github.com/slimgroup/JUDI.jl/blob/master/examples/notebooks/05_custom_misfit.ipynb

The envelope misfit works too so not sure what you did here that made it error.
